pkg/hub/controllers/cluster/reconciler.go

```diff
@@ -458,3 +489,50 @@ func (r *Reconciler) InjectClient(c client.Client) error {
 	r.Client = c
 	return nil
 }
+
+func (r *Reconciler) handleClusterDeletion(cluster *hubv1alpha1.Cluster, ctx context.Context, req reconcile.Request) (bool, reconcile.Result, error) {
+	log := logger.FromContext(ctx)
+
+	if cluster.ObjectMeta.DeletionTimestamp.IsZero() {
+		// The object is not being deleted, so if it does not have our finalizer,
+		// then lets add the finalizer and update the object. This is equivalent
+		// registering our finalizer.
+		if !controllerutil.ContainsFinalizer(cluster, clusterDeregisterFinalizer) {
+			controllerutil.AddFinalizer(cluster, clusterDeregisterFinalizer)
+			if err := r.Update(ctx, cluster); err != nil {
+				return true, reconcile.Result{}, err
+			}
+		}
+	} else {
+		// The object is being deleted
+		if controllerutil.ContainsFinalizer(cluster, clusterDeregisterFinalizer) && !cluster.Status.IsDeregisterInProgress && retryAttempts <= MAX_CLUSTER_DEREGISTRATION_ATTEMPTS {
```

Comment:
MAX_CLUSTER_DEREGISTRATION_ATTEMPTS is set to 3. But here its '<=' which means it will attempt 4 times. Should this run 3 or 4 attempts? If 3 then it should be '<', not '<='.